crates/notion-core/src/tool.rs

```diff
@@ -139,8 +171,43 @@ impl Tool for Script {
 }
 
 impl Tool for Binary {
-    fn new(_session: &mut Session) -> Fallible<Self> {
-        unimplemented!()
+    fn new(session: &mut Session) -> Fallible<Self> {
+        session.add_event_start(ActivityKind::Binary);
+
+        let mut args = args_os();
+        let exe = arg0(&mut args)?;
+        let current_path = var_os("PATH").unwrap_or(OsString::new());
+
+        if let Some(project) = session.project() {
+            if project.has_local_bin(&exe)? {
+                // local binary
+                // use the full path to the binary
+                let mut path_to_bin = project.local_bin_dir();
+                path_to_bin.push(&exe);
+                return Ok(Self::from_components(
+                    &path_to_bin.as_os_str(),
+                    args,
+                    &current_path,
+                ));
+            }
+        }
+
+        if let Some(version) = session.current_node()? {
+            // globally installed binary for this version of node
+            // use the full path to the binary
+            let mut third_p_bin_dir = path::node_version_3p_bin_dir(&version.to_string())?;
+            third_p_bin_dir.push(&exe);
+            return Ok(Self::from_components(
+                &third_p_bin_dir.as_os_str(),
+                args,
+                &current_path,
+            ));
+        };
+
+        // globally installed system binary
+        // not in a notion project, so use the system binary (remove notion shims and bins)
```

Comment:
I think this comment is wrong, but I'm not 100% sure. I've been studying this code for a while and I think we may have a misunderstanding. Unless I'm mistaken, the method `session.project()` returns `Some` if we're in a _Node package_, not if we're in a _Notion project_.

So that would mean that your code is doing the right thing, it's just the comment and the description of this PR that are wrong. The "priority algorithm" goes:

1. If the executable is installed locally, and a direct dependency of the current Node project, then we execute that. We only execute direct dependencies because npm and yarn install executables for everything installed in node_modules/, not just direct dependencies. Exposing all of those on the path would be a security risk as well as confusing to debug.
2. If we're in a Node project, but the executable is not local, we execute the globally-installed executable for the configured version of node.
3. If we are not in a Node project, we run the executable installed globally with the system node.

Reply:
I updated the comments for this in the code, also going to update the description in this PR